src/runtime/hexagon/hexagon/hexagon_buffer.cc

```diff
@@ -30,67 +34,135 @@ namespace tvm {
+  #else
+    data_ = malloc(nbytes);
+  #endif
   }
-  allocations_.push_back(ptr);
-  SetStorageScope(scope);
+  ~VTCMAllocation() {
+  #ifdef BUILD_FOR_HEXAGON
+    // HEXAGON_PRINT(ALWAYS, "~VTCMAllocation() - Context ID: %u, VTCM ptr: %p", context_id_, data_);
+    // TODO: Need to handle the else case(s) here
+    if (context_id_ && data_) {
+      HEXAGON_SAFE_CALL(HAP_compute_res_release(context_id_));
+      data_ = nullptr;
+    }
+  #else
+    free(data_);
+  #endif
+  }
+  #ifdef BUILD_FOR_HEXAGON
+  unsigned int context_id_{0};
+  #endif
+};
+
+template <HexagonBuffer::StorageScope S>
+std::unique_ptr<Allocation> Allocator(size_t nbytes, size_t alignment);
+
+template <>
+std::unique_ptr<Allocation> Allocator<HexagonBuffer::StorageScope::kDDR>(size_t nbytes,
+                                                                         size_t alignment) {
+  return std::make_unique<DDRAllocation>(nbytes, alignment);;
 }
 
-HexagonBuffer::HexagonBuffer(void* data, Optional<String> scope) : managed_{false} {
+template <>
+std::unique_ptr<Allocation> Allocator<HexagonBuffer::StorageScope::kVTCM>(size_t nbytes,
+                                                                          size_t alignment) {
+  return std::make_unique<VTCMAllocation>(nbytes, alignment);
+}
+
+HexagonBuffer::HexagonBuffer(size_t nbytes, size_t alignment, Optional<String> scope) : ndim_(1), nbytes_(nbytes) {
   SetStorageScope(scope);
-  allocations_.push_back(data);
+
+  std::unique_ptr<Allocation> alloca = nullptr;
+  if (GetStorageScope() == StorageScope::kDDR) {
+    alloca = Allocator<StorageScope::kDDR>(nbytes, alignment);
+  }
+  else if (GetStorageScope() == StorageScope::kVTCM) {
+    alloca = Allocator<StorageScope::kVTCM>(nbytes, alignment);
+  }
+  CHECK(alloca != nullptr);
+  allocations_.push_back(alloca->data_);
+  managed_allocations_.push_back(std::move(alloca));
 }
 
-HexagonBuffer::~HexagonBuffer() {
-  if (managed_) {
-    for (auto& ptr : allocations_) {
-      free(ptr);
+HexagonBuffer::HexagonBuffer(size_t ndim, size_t nbytes, size_t alignment, Optional<String> scope) : ndim_(ndim), nbytes_(ndim * nbytes) {
+  SetStorageScope(scope);
+  for (size_t i = 0; i < ndim; ++i) {
+    std::unique_ptr<Allocation> alloca = nullptr;
+    if (GetStorageScope() == StorageScope::kDDR) {
+      alloca = Allocator<StorageScope::kDDR>(nbytes, alignment);
+    }
+    else if (GetStorageScope() == StorageScope::kVTCM) {
+      alloca = Allocator<StorageScope::kVTCM>(nbytes, alignment);
     }
+    CHECK(alloca != nullptr);
+    allocations_.push_back(alloca->data_);
+    managed_allocations_.push_back(std::move(alloca));
   }
 }
 
-HexagonBuffer::HexagonBuffer(HexagonBuffer&& other)
-    : allocations_(other.allocations_),
-      managed_(other.managed_),
-      storage_scope_(other.storage_scope_) {
-  other.allocations_.clear();
-  other.managed_ = false;
-  other.storage_scope_ = StorageScope::kDDR;
+// TODO: add ExternalAllocation class?
```

Comment:
The separate tracking / management of `allocations_` and `managed_allocations_` is bothersome and could be solved by an `ExternalAllocation` class similar to what is done with VTCM / DDR.  It may be that `ExternalAllocation` is a separate `DDRAllocation` where the data pointer is provided rather than allocated given that the `Allocator` is templated based on storage scope.